Feature: Logframe page enhancements

MalloryLeeWong opened the pull request:
### Changes
This PR enhances the Logframe page so the user can view, add, and edit all the Impact, Outcome, and Output details, including Outcome indicators, Output indicators, and Output Updates.

- https://github.com/Blue-Marine-Foundation/Maerl/issues/154 (This demo feedback overrides some of issue #141 below)
- https://github.com/Blue-Marine-Foundation/Maerl/issues/141
- https://github.com/Blue-Marine-Foundation/Maerl/issues/146 (implements Part 1 to update the form label copy, not Part 2 which will be handled in the future)

### Screenshots

#### Logframe page
<img width="1665" alt="Screenshot 2025-02-18 at 12 58 09 PM" src="https://github.com/user-attachments/assets/0f83ed98-5fec-4096-8b88-c9d8295a351e" />

<img width="1650" alt="Screenshot 2025-02-18 at 12 58 54 PM" src="https://github.com/user-attachments/assets/92bfae17-3665-44ca-9664-37cb3a9e9561" />

<img width="1648" alt="Screenshot 2025-02-18 at 12 59 10 PM" src="https://github.com/user-attachments/assets/06a6237a-e7b6-4263-9160-63d9b7ce2c8f" />

<img width="1486" alt="Screenshot 2025-02-18 at 12 59 36 PM" src="https://github.com/user-attachments/assets/0bc33049-d472-443d-a1cd-5a086a159e34" />

<img width="1395" alt="Screenshot 2025-02-18 at 12 59 57 PM" src="https://github.com/user-attachments/assets/f589d217-fdd9-4ee2-ac19-e10ff75f88c5" />

<img width="1389" alt="Screenshot 2025-02-18 at 1 00 10 PM" src="https://github.com/user-attachments/assets/fea665f6-1308-4b8a-abd2-2c4e7a1c164c" />

#### Copy update for issue #146 
![Screenshot 2025-02-18 at 1 01 51 PM](https://github.com/user-attachments/assets/6e228dbe-6f5d-400e-a31c-3c273e3ef423)

MalloryLeeWong commented:
@samjamead Created this new draft PR for the remaining Logframe enhancements I'm implementing on the Output detail page. I'm almost done. The main tasks left are to implement the add/edit output indicator Updates and Activities buttons and forms.

Would love to get any thoughts you have on the new direction of the UI layout so far. Please find screenshots attached.

I am interested in showing this to Sophie during the demo to get her feedback even if the edit functionality isnt' all done yet, then aim to finish up implementing this view this week. I can let her know that this is subject to your technical and UI design review as well and it's just an initial rendition with my ideas so far.

samjamead commented:
Heya @MalloryLeeWong thanks for your progress! And thanks for prompting me to take a look at the work in progress. 

Overall I like the direction so far and I like how you've brought continuity between these pages and the ToC / Logframe pages – the Output pages look like an obvious continuation which is great for the user. 

Using tabs to navigate between Updates and Activities is great too, I like that approach a lot. 

I think my advice is to move away from the nested heirarchy layout that I introduced in the current production version. I.e. I think it makes sense for the Updates and Activities to live outside of their parent table row. The current layout has worked to a point, but only in the sense that I have cheated and bundled all the updates for the whole output together. 

Now that we're going to move towards updates (and activities) being listed under each specific Output Indicator, I think nesting makes less sense. 

I'm not 100% sure on how I would lay this out in the UI but let me sleep on it and come up with some ideas. 

One thing I do know is that Sophie and Suneha like having the updates listed as tables (and they're looking to add more columns too in future), so I expect they'd want to retain that view. 

### UI libraries

Apologies, I should have mentioned this during our kickoff call! Actually, so far the codebase does use a single UI library – all of the components in `@/components/ui`, and a few others scattered throughout, come from ShadCN! 

I appreciate that this is not immediately obvious because ShadCN/UI is not installed as a dependency in the normal way. In fact ShadCN says himself in the docs: 

> This is not a component library. It is how you build your component library.

Things do work slightly differently in that you add components to the repo via copy/paste (or I use the command line shortcuts), and then you customise them to fit within the project component library (for me that's generally `@/components/ui` unless they're a bit more complex and likely to be single-use, as you've spotted with, e.g., the calendar date picker). 

In terms of next steps, my recommendation is to have a quick look at the ShadCN docs and see if you can trace things back to the Maerl component library (`popover` or `dialogue` are good ones because they're used in a handful of places throughout the app). If you like it, then great, we can stick with it. 

On the other hand, if you don't find it intuitive, I'm happy to use Headless too – that's kind of the beauty of ShadCN – because the components are copied into your repo, rather than imported, it's not such a big deal to also use another component library if you need to.

MalloryLeeWong commented:
@samjamead Thanks for your comments! 

Cool, didn't know that's how ShadCN works! I will look at incorporating the ShadCN components for this PR instead of using Headless UI.

For the output indicator updates and activities, that's helpful to know that the team has a preference for keeping these in table format and that more columns could be added in the future. From experimenting so far, I have found that it looks a bit overwhelming to have a nested table under each parent table row, so that's how I ended up on the current tab list view idea. I will keep brainstorming how to keep the updates and activities as tables. Appreciate any ideas you have as well!

MalloryLeeWong commented:
@samjamead This PR is ready for your review. 

As discussed, this excludes 2 features: the Output Activities list and displaying the new Output Actual Value column since we will be handling those separately as fast followers once the db updates are complete.

Also as an FYI, I removed the Headless UI components that I had introduced in my earlier draft since the demo feedback resulted in those components no longer being needed anyways. Also liking continuing to use shadcn instead.

samjamead commented:
Brill thanks @MalloryLeeWong I have two hours tomorrow to work on all things Maerl so I will do PR reviews, a sweep of the issues and any outstanding database changes 👍

MalloryLeeWong commented:
> Brill thanks @MalloryLeeWong I have two hours tomorrow to work on all things Maerl so I will do PR reviews, a sweep of the issues and any outstanding database changes 👍

Sounds great, thanks @samjamead! I'll plan to check on this to address PR feedback tomorrow on Wednesday evening or Thursday  afternoon my time.

MalloryLeeWong commented:
> Hey @MalloryLeeWong just wrapping up on a good look over everything here. As per last time no real comments on your code! I think we have a similar style and I like how clean and minimal your approach is.
> 
> **Extra thanks for:**
> 
> * Tightening up the spacing in a few places
> * The collapsible / accordions are a great addition and help navigate the logframe
> * New components (specific cards) are logically added and nicely implemented
> 
> **Changes:**
> 
> * I made one very small change: to display the relevant impact indicator unit alongside an update.
> 
> Clear to merge!
> 
> My one question mark – I understand that the logframe page is now based on feedback from Soph and Suneha so no real arguments there. Based on the testing I did with a small group of PMs around November 2023, I can see maybe a few usability issues springing up from having everything in one place.
> 
> **Recommendation** – Let's merge your excellent PR and get this functionality live, but perhaps in March we can look to iterate on this quickly and revisit the UX and the UI. In particular I'd like to tighten up the workflow for PMs adding updates to their projects – we'll need to do some user testing to figure out the best way to do this I think.

Thanks @samjamead, appreciate your feedback and teh change you made to display the units - great idea! Agreed with your suggested approach to do further user testing on the new workflow and iterating on the UX/UI as needed in March.
